File: scala/stackful_coroutine/ThreadDemo.scala
/**
 * ThreadDemo.scala
 *
 * Paradigm: Stackful Coroutine (via Thread + BlockingQueue)
 *
 * Scala/Java threads are stackful: each thread has its own call stack.
 * When a thread blocks on queue.take(), the full call stack is preserved.
 * This is the same model as Go goroutines and Lua coroutines.
 *
 * This demo simulates a worker that can "suspend" at any call depth
 * and be "resumed" later — the defining property of stackful coroutines.
 *
 * Compile:  scalac -d build stackful_coroutine/ThreadDemo.scala
 * Run:      scala -classpath build ThreadDemo
 */

import java.util.concurrent.{LinkedBlockingQueue, BlockingQueue}

object ThreadDemo {

  /** A "suspendable" worker simulation. */
  class SuspendableWorker(name: String, workItems: Seq[Int]) {
    private val mailbox = new LinkedBlockingQueue[String]()

    private val thread = new Thread {
      override def run(): Unit = {
        var depth = 0
        for (item <- workItems) {
          println("  [" + name + "] step " + item + " (depth=" + depth + ")")
          Thread.sleep(10)

          // Simulate nested call that can "suspend"
          val result = innerProcess(item, depth)
          depth = result

          // "Suspend" — thread blocks here, stack is preserved
          try {
            mailbox.put(name + " completed item " + item)
          } catch {
            case _: InterruptedException => // exit
          }
        }
        println("  [" + name + "] DONE")
      }
    }

    thread.start()

    def waitForNextResult(): String = mailbox.take()

    def join(): Unit = thread.join()
  }

  private def innerProcess(item: Int, depth: Int): Int = {
    // Simulate deep call chain (stackful = we can suspend from any depth)
    if (item % 2 == 0) {
      evenHandler(item, depth)
    } else {
      oddHandler(item, depth)
    }
  }

  private def evenHandler(item: Int, depth: Int): Int = {
    println("    (evenHandler depth=" + depth + ")")
    depth + 2
  }

  private def oddHandler(item: Int, depth: Int): Int = {
    println("    (oddHandler depth=" + depth + ")")
    depth + 1
  }

  def main(args: Array[String]): Unit = {
    println("=== Stackful Coroutine via Thread ===\n")

    val worker1 = new SuspendableWorker("W1", Seq(1, 2, 3))
    val worker2 = new SuspendableWorker("W2", Seq(4, 5, 6))

    // Main thread collects results (like a scheduler)
    for (_ <- 0 until 3) {
      println("  [Main] result: " + worker1.waitForNextResult())
      println("  [Main] result: " + worker2.waitForNextResult())
    }

    worker1.join()
    worker2.join()

    println("\n=== Key insight ===")
    println("Threads are stackful: each thread has a full call stack.")
    println("When a thread blocks (e.g., on queue.take()), the entire")
    println("stack is preserved. This is the same model as Go goroutines")
    println("(stackful) vs Python generators (stackless).")
  }
}

File: kotlin/stackful_coroutine/ThreadDemo.kt
/**
 * ThreadDemo.kt
 *
 * Paradigm: Stackful Coroutine (via threads)
 *
 * Kotlin coroutines are stackless by default. This demo uses
 * traditional threads to demonstrate stackful semantics:
 * each thread has its own full call stack that is preserved
 * when the thread blocks.
 *
 * Compare with: Lua coroutines (stackful), Go goroutines (stackful)
 *
 * Compile: kotlinc -d out stackful_coroutine/ThreadDemo.kt
 * Run:     java -cp /usr/share/java/kotlin-stdlib.jar:out ThreadDemoKt
 */

import java.util.concurrent.LinkedBlockingQueue

fun main(args: Array<String>) {
    println("=== Stackful Coroutine via Threads ===\n")

    val results = LinkedBlockingQueue<String>()

    // Worker 1 — simulates a stackful coroutine with a full call stack
    val thread1 = Thread {
        worker("A", listOf(1, 2, 3), results)
    }

    // Worker 2
    val thread2 = Thread {
        worker("B", listOf(4, 5, 6), results)
    }

    thread1.start()
    thread2.start()

    // Collect results (like a scheduler)
    for (i in 1..6) {
        println("  [Main] got: " + results.take())
    }

    thread1.join()
    thread2.join()

    println("\n=== Key insight ===")
    println("Threads are stackful: each thread has a full call stack.")
    println("When a thread blocks on LinkedBlockingQueue.take(), the")
    println("entire call stack is preserved. This is the same model")
    println("as Go goroutines and Lua coroutines.")
}

/** A worker with nested calls that can "suspend" at any depth. */
fun worker(name: String, items: List<Int>, results: LinkedBlockingQueue<String>) {
    for (item in items) {
        println("  [$name] processing item $item")
        val result = deepProcess(name, item)
        results.put(result)
    }
    println("  [$name] DONE")
}

fun deepProcess(name: String, item: Int): String {
    return if (item % 2 == 0) {
        evenHandler(name, item)
    } else {
        oddHandler(name, item)
    }
}

fun evenHandler(name: String, item: Int): String {
    Thread.sleep(10)
    return "$name processed even item $item"
}

fun oddHandler(name: String, item: Int): String {
    Thread.sleep(10)
    return "$name processed odd item $item"
}
